amaranth/vendor/gowin.py

```diff
@@ -0,0 +1,556 @@
+from abc import abstractproperty
+from fractions import Fraction
+
+from ..hdl import *
+from ..lib.cdc import ResetSynchronizer
+from ..build import *
+
+# Acknowledgments:
+#   Parts of this file originate from https://github.com/tcjie/Gowin
+
+
+__all__ = ["GowinPlatform"]
+
+
+class GowinPlatform(TemplatedPlatform):
+    """
+    .. rubric:: Apicula toolchain
+
+    Required tools:
+        * ``yosys``
+        * ``nextpnr-gowin``
+        * ``gowin_pack``
+
+    The environment is populated by running the script specified in the environment variable
+    ``AMARANTH_ENV_APICULA``, if present.
+
+    Build products:
+        * ``{{name}}.fs``: binary bitstream.
+
+    .. rubric:: Gowin toolchain
+
+    Required tools:
+        * ``gw_sh``
+
+    The environment is populated by running the script specified in the environment variable
+    ``AMARANTH_ENV_GOWIN``, if present.
+
+    Build products:
+        * ``{{name}}.fs``: binary bitstream.
+    """
+
+    toolchain = None # selected when creating platform
+
+    device    = abstractproperty()
```

Comment:
Setting all of these properties individually is quite confusing: (from Tang Nano)
```
    # GW1N-LV1QN48C6/I5
    device      = "GW1N-1"
    series      = "GW1N"
    size        = "1"
    subseries   = ""
    voltage     = "LV"
    package     = "QN48"
    speed       = "C6/I5"
```

Can we do better here?

Also, and this is related, I'm not entirely sure why we have both `_device_name` and `_part`; the naming makes it especially difficult to understand what's going on.

Reply:
Their part/family naming scheme is bonkers. Also the option names of the tools are not quite straightforward. 

We could just have the long part name as the sole property and then try to derive everything that's necessary inside of GowinPlatform.

`_part` and `_device_name` are different concatenations of the part name constituents and needed by the different tools in exactly this format.

Reply:
> We could just have the long part name as the sole property and then try to derive everything that's necessary inside of GowinPlatform.

This is what `XilinxPlatform` does. If you're up for that I think that's ideal, but considering how bonkers it is I don't feel like I can _insist_ on this.

> `_part` and `_device_name` are different concatenations of the part name constituents and needed by the different tools in exactly this format.

If you're _not_ up for it then I propose workshopping a better naming for these two, for any unlucky soul who has to work on this after us.